### PR TITLE
Cover blocked response-context TCK semantics

### PR DESCRIPTION
**Pull request by danielflower**
## What changed

Adds direct and integration coverage for the Jakarta REST `ContainerResponseContext` behavior hidden by the TCK resource field-injection prerequisite.

The audit exposed and fixes three real bugs:

- `getHeaderString` now converts and comma-joins every value, while distinguishing a missing header (`null`) from a present header with no value (`""`).
- `getStringHeaders()` is now a live, converting view of the object-valued headers map. Changes through `getHeaders()` are immediately visible, and existing string-map mutation behavior now writes through to the response headers.
- Replacing the response entity output stream suppresses automatic raw-entity `Content-Length` calculation and the runtime closes the replacement stream, so transformed or buffered bodies complete safely. An explicit length supplied by application code is preserved.

The tests also cover entity metadata, annotations, streams, status mutation, headers, cookies, links, null/default values, and the standard response metadata accessors exercised by the 57 blocked response-context cases.

## TDD evidence

- `getHeaderStringJoinsEveryHeaderValue` failed with only `first` before the fix.
- `stringHeadersAreALiveViewOfObjectHeaders` failed because a captured string-header map remained an empty snapshot after object-header mutations.
- The response-filter integration test now verifies that `getStringHeaders().putSingle(...)` changes the emitted HTTP header.
- `responseFiltersCanReplaceTheEntityOutputStream` failed with an HTTP/2 stream reset because Mu declared 4 bytes then wrote 11.
- The buffered-stream variant produced an incomplete body until the runtime closed the replacement stream.
- Fresh review caught and corrected the empty-header and explicit-`Content-Length` edge cases before merge.

## Validation

All validation used Java 11:

- `mvn -q -Dtest=JaxRSResponseTest,FilterTest,HeadTest,WriterInterceptorTest test`
- `mvn -q test`
- Focused Jakarta REST 3.1 harness: 57 `BLOCKED` by `FIELD_INJECTION_PER_REQUEST`, 0 unexpected results
- `git diff --check`

### CODE REVIEW

**Comment by danielflower** [comment]
@codex review

File: src/main/java/io/muserver/rest/RestHandler.java [hunk 1]
```diff
@@ -279,6 +279,8 @@ private void sendResponse(int nestingLevel, JaxRSRequest requestContext, MuRespo
                     if (jaxRSResponse.hasEntity()) {
                         jaxRSResponse.executeInterceptors(writerInterceptors); // run the interceptors
                     }
+                    boolean entityStreamReplaced = requestContext.getMuMethod() != Method.HEAD
+                        && jaxRSResponse.getOutputStream() != out;
```
[the remark this answers is not in the document]
Comment (danielflower) [reply src/main/java/io/muserver/rest/RestHandler.java]: Agreed. I added a HEAD request to the wrapping-filter regression and changed detection to compare with the entity stream initially assigned to the request. This preserves Content-Length for ordinary HEAD responses while omitting the unknowable transformed length when a filter wraps the HEAD stream. `HeadTest`, `FilterTest`, and `JaxRSResponseTest` all pass on Java 17.

**Comment by danielflower** [comment]
@codex review

File: src/main/java/io/muserver/rest/RestHandler.java [hunk 2]
```diff
@@ -302,7 +305,9 @@ private void sendResponse(int nestingLevel, JaxRSRequest requestContext, MuRespo
                         Type entityGenericType = jaxRSResponse.getEntityType();
                         MessageBodyWriter messageBodyWriter = entityProviders.selectWriter(entityType, entityGenericType, writerAnnontations, responseMediaType);
 
-                        if (entityProviders.isBuiltInWriter(messageBodyWriter)) {
+                        if (entityStreamReplaced) {
+                            jaxRSResponse.getHeaders().remove(HttpHeaders.CONTENT_LENGTH);
```
[the remark this answers is not in the document]
Comment (danielflower) [reply src/main/java/io/muserver/rest/RestHandler.java]: Confirmed by changing the end-to-end filter regression to wrap the original stream in `BufferedOutputStream`: without lifecycle handling the body is not completed. After a successful provider write, RestHandler now closes only the replacement stream before completing the original lazy stream. `HeadTest`, `FilterTest`, `JaxRSResponseTest`, and `WriterInterceptorTest` pass on Java 17.

**Comment by danielflower** [comment]
@codex review

File: src/main/java/io/muserver/rest/JaxRSResponse.java [hunk 3]
```diff
@@ -401,7 +401,10 @@ static MultivaluedMap<String, Object> muHeadersToJaxObj(Headers headers) {
 
     @Override
     public String getHeaderString(String name) {
-        return headerValueToString(headers.getFirst(name));
+        List<Object> values = headers.get(name);
+        return values == null ? null : values.stream()
+            .map(JaxRSResponse::headerValueToString)
+            .collect(Collectors.joining(","));
```
[the remark this answers is not in the document]
Comment (danielflower) [reply src/main/java/io/muserver/rest/JaxRSResponse.java:402]: Confirmed. I added `getHeaderStringTreatsAnEmptyValueListAsAbsent` and now return null for both a missing mapping and an empty mapped list, while retaining comma joining for non-empty lists. The focused response/filter/HEAD/interceptor regression suites pass on Java 17.

**Comment by danielflower** [comment]
@codex review

**Comment by danielflower** [comment]
@codex review

**Comment by danielflower** [comment]
Follow-up map-contract fix is now pushed at `3ae4d7b8`. @codex review

File: src/main/java/io/muserver/rest/RestHandler.java [hunk 4]
```diff
@@ -314,8 +317,12 @@ private void sendResponse(int nestingLevel, JaxRSRequest requestContext, MuRespo
                         applyDefaultCharset(jaxRSResponse);
 
                         try {
+                            OutputStream responseEntityStream = jaxRSResponse.getOutputStream();
                             messageBodyWriter.writeTo(jaxRSResponse.getEntity(), jaxRSResponse.getType(), jaxRSResponse.getGenericType(), writerAnnontations,
-                                jaxRSResponse.getMediaType(), jaxRSResponse.getHeaders(), jaxRSResponse.getOutputStream());
+                                jaxRSResponse.getMediaType(), jaxRSResponse.getHeaders(), responseEntityStream);
+                            if (entityStreamReplaced) {
+                                responseEntityStream.close();
+                            }
```
[the remark this answers is not in the document]
Comment (danielflower) [reply src/main/java/io/muserver/rest/RestHandler.java]: Confirmed with a new end-to-end regression, `responseFiltersCloseReplacementEntityStreamsWhenWritingFails`: a custom writer throws before writing, the client receives the fallback 500, and the replacement stream was previously left open. Commit `b0cb11b9` now manages only the replacement stream with try-with-resources, so it is closed on success and failure while the original writer exception remains primary if close also fails. The focused response/filter/HEAD/interceptor suites and the full Java 11 suite pass.

File: src/main/java/io/muserver/rest/StringHeadersMap.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+                            return entries.hasNext();
+                        }
+
+                        @Override
+                        public Entry<String, List<String>> next() {
+                            Entry<String, List<Object>> entry = entries.next();
+                            return new SimpleEntry<String, List<String>>(entry.getKey(), null) {
+                                @Override
+                                public List<String> getValue() {
+                                    return asStringList(entry.getValue());
```
[the remark this answers is not in the document]
Comment (danielflower) [reply src/main/java/io/muserver/rest/StringHeadersMap.java]: Confirmed against the reviewed commit, and already fixed in follow-up `3ae4d7b8` by replacing `SimpleEntry` with an entry whose equality and hash code consult the live converted value. Commit `b0cb11b9` adds `stringHeaderEntriesObeyTheMapEqualityAndHashCodeContracts`, which proves symmetric entry equality plus map and entry hash-code parity with an ordinary map. It passes in the focused and full Java 11 suites.

**Comment by danielflower** [comment]
Both review findings are now covered by explicit regressions, all review threads are resolved, and CI is green on Java 11/17/21/25 at `b0cb11b9`. @codex review

**Comment by danielflower** [comment]
@codex review

**Comment by danielflower** [comment]
Added manual `GZIPOutputStream` response-filter coverage at `704a82fc`. This exposed and fixes early response commitment from the gzip header written by the wrapper constructor; bytes emitted during filter/interceptor setup are now deferred until entity writing, and automatic `Content-Length` is skipped when interceptors already emitted bytes. Verified over HTTP/1.1 and HTTP/2 plus focused response/filter/HEAD/interceptor suites. @codex review

**Comment by danielflower** [comment]
Added the complementary gzipped request-stream test at `73e1b580`: a request filter wraps the encoded entity in `GZIPInputStream`, removes stale `Content-Encoding`/`Content-Length`, preserves `Content-Type`, and the resource receives the original text. The test exposed and fixes `InputStream.read()` returning signed bytes for values above `0x7f`. Verified over HTTP/1.1 and HTTP/2, plus the complete request-stream adapter and filter suites. @codex review

File: src/main/java/io/muserver/rest/RestHandler.java [hunk 6]
```diff
@@ -314,9 +317,13 @@ private void sendResponse(int nestingLevel, JaxRSRequest requestContext, MuRespo
                         applyDefaultCharset(jaxRSResponse);
 
                         try {
-                            messageBodyWriter.writeTo(jaxRSResponse.getEntity(), jaxRSResponse.getType(), jaxRSResponse.getGenericType(), writerAnnontations,
-                                jaxRSResponse.getMediaType(), jaxRSResponse.getHeaders(), jaxRSResponse.getOutputStream());
-                            out.prepare();
+                            out.releaseWrites();
+                            OutputStream responseEntityStream = jaxRSResponse.getOutputStream();
+                            try (OutputStream replacementStream = entityStreamReplaced ? responseEntityStream : null) {
```
[the remark this answers is not in the document]
Comment (danielflower) [reply src/main/java/io/muserver/rest/RestHandler.java]: Confirmed. The gzip failure regression now uses `GZIPOutputStream` itself: the writer throws before producing bytes, the wrapper is closed, and the client must still receive 500. In `46cd4f5c`, `discardUncommittedWrites()` is invoked before closing the replacement, so its footer is discarded while the original writer exception remains primary. The regression passes on HTTP/1.1 and HTTP/2.

File: src/main/java/io/muserver/rest/RestHandler.java [hunk 7]
```diff
@@ -314,9 +317,13 @@ private void sendResponse(int nestingLevel, JaxRSRequest requestContext, MuRespo
                         applyDefaultCharset(jaxRSResponse);
 
                         try {
-                            messageBodyWriter.writeTo(jaxRSResponse.getEntity(), jaxRSResponse.getType(), jaxRSResponse.getGenericType(), writerAnnontations,
-                                jaxRSResponse.getMediaType(), jaxRSResponse.getHeaders(), jaxRSResponse.getOutputStream());
-                            out.prepare();
+                            out.releaseWrites();
```
[the remark this answers is not in the document]
Comment (danielflower) [reply src/main/java/io/muserver/rest/RestHandler.java]: Confirmed. `responseFiltersCanWriteLargeBodiesWithoutAnEntity` writes more than twice the deferral bound directly from a response filter while leaving the entity null; it verifies the filter-selected 201 status and every body byte. The null-entity path now closes replacements and finishes any direct output instead of declaring `Content-Length: 0`.

File: src/main/java/io/muserver/rest/RestHandler.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (danielflower) [reply src/main/java/io/muserver/rest/RestHandler.java]: Confirmed; this duplicates the earlier gzip failure report and is covered by the same explicit `GZIPOutputStream` failure regression. Commit `46cd4f5c` discards only while the live output is still uncommitted, then closes the wrapper. If output was already streamed it is finalized because an HTTP error can no longer replace a committed response.

File: src/main/java/io/muserver/rest/LazyAccessOutputStream.java [hunk 8]
```diff
@@ -14,11 +15,18 @@ class LazyAccessOutputStream extends OutputStream {
...
     private OutputStream os;
     private boolean prepared;
+    private boolean writesReleased;
+    private ByteArrayOutputStream deferredBytes = new ByteArrayOutputStream();
 
-    private OutputStream out() {
+    private OutputStream out() throws IOException {
+        if (!writesReleased) {
+            return deferredBytes;
+        }
```
[the remark this answers is not in the document]
Comment (danielflower) [reply src/main/java/io/muserver/rest/LazyAccessOutputStream.java]: Confirmed and redesigned in `46cd4f5c`. Deferral is now strictly capped at 8192 bytes, enough for wrapper preambles such as gzip. The 8193rd byte releases the buffer and streams all subsequent output. `LazyAccessOutputStreamTest.streamsInsteadOfBufferingBeyondTheDeferredByteLimit` proves the exact boundary, and the large entity-less response-filter integration test proves end-to-end streaming without data loss.

**Comment by danielflower** [comment]
Addressed all four new findings in `46cd4f5c`: pre-commit buffering is hard-capped at 8 KiB with exact-boundary coverage; gzip close-time bytes are discarded when a writer fails before commitment; and large entity-less filter output streams successfully. All threads are resolved and the full Java 11 suite passes locally apart from the explicitly excluded fixed-port test whose port is occupied by an unrelated local process. @codex review

File: src/main/java/io/muserver/rest/RestHandler.java [hunk 9]
```diff
@@ -314,9 +329,29 @@ private void sendResponse(int nestingLevel, JaxRSRequest requestContext, MuRespo
...
+                            } catch (Exception e) {
+                                writeFailure = e;
+                                out.discardUncommittedWrites();
+                            }
+                            if (entityStreamReplaced) {
+                                try {
+                                    responseEntityStream.close();
+                                } catch (Exception closeFailure) {
+                                    if (writeFailure == null) {
+                                        throw closeFailure;
```
[the remark this answers is not in the document]
Comment (danielflower) [reply src/main/java/io/muserver/rest/RestHandler.java]: Confirmed with two end-to-end regressions. Before the fix, both an ordinary entity and an entity-less response returned 200 after a buffered replacement flushed during close and then threw. Commit `04381b75` now re-defers still-uncommitted writes while closing a replacement, discards them if close fails, and commits them with `finish()` only after close succeeds. `LazyAccessOutputStream.close()` no longer implicitly commits deferred bytes. The deferral remains hard-capped at 8 KiB; output beyond that is still streamed and cannot be rolled back. The focused lifecycle/gzip/large-output tests pass, and the Java 11 suite has 943 passes with only the known unrelated fixed-port collision.

**Comment by danielflower** [comment]
The replacement-close failure finding is fixed in `04381b75` with explicit regressions for both entity and entity-less responses. The thread is resolved and CI is green on Java 11/17/21/25. @codex review

**Comment by danielflower** [comment]
Simplified the response-stream implementation in `03d6888d` after revisiting the design tradeoff. `LazyAccessOutputStream` is restored exactly to its original implementation: no deferred byte buffer, rollback state, size threshold, or special close semantics (241 lines removed, 12 added overall). The real gzip failure is handled by applying the filter-selected status in the existing first-write callback and avoiding a second status mutation after commitment. Replacement streams are still closed by the runtime, including on writer failure; once a wrapper emits bytes, normal streaming semantics apply and the response is committed. The gzip regression now also verifies a filter-selected 201 status. The full Java 11 suite passes locally (938 tests, excluding only the unrelated fixed-port test), and CI is green on Java 11/17/21/25. @codex review
